File: backend/app/models/message.py
"""
Message database model.
"""
from sqlalchemy import Column, String, Boolean, DateTime, ForeignKey, func, Text
from sqlalchemy.dialects.postgresql import UUID
import uuid
from app.database import Base


class Message(Base):
    """Unified message model for all communication sources."""
    
    __tablename__ = "messages"
    
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False, index=True)
    listing_id = Column(UUID(as_uuid=True), ForeignKey("listings.id"), nullable=True, index=True)
    
    # Message source and identification
    source = Column(String(50), nullable=False)  # email, ebay, facebook, etc.
    external_id = Column(String(255), nullable=True)  # ID from source platform
    
    # Sender information
    sender_name = Column(String(255), nullable=True)
    sender_email = Column(String(255), nullable=True)
    
    # Message content
    subject = Column(String(500), nullable=True)
    body = Column(Text, nullable=False)
    
    # Status and AI
    is_read = Column(Boolean, default=False)
    ai_draft_response = Column(Text, nullable=True)
    
    # Threading
    thread_id = Column(String(255), nullable=True, index=True)
    
    # Timestamps
    received_at = Column(DateTime(timezone=True), nullable=False)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    
    def __repr__(self):
        return f"<Message from {self.sender_name} via {self.source}>"

File: market_forge_app/lib/models/message.dart
enum MessageSource {
  email,
  ebay,
  facebook,
  craigslist,
  mercari,
  etsy,
  amazon,
}

class Message {
  final String id;
  final MessageSource source;
  final String senderName;
  final String senderEmail;
  final String subject;
  final String body;
  final String? listingId;
  final DateTime timestamp;
  final bool isRead;
  final String? aiDraftResponse;

  Message({
    required this.id,
    required this.source,
    required this.senderName,
    required this.senderEmail,
    required this.subject,
    required this.body,
    this.listingId,
    required this.timestamp,
    this.isRead = false,
    this.aiDraftResponse,
  });

  factory Message.fromJson(Map<String, dynamic> json) {
    return Message(
      id: json['id'] as String,
      source: _parseSource(json['source'] as String),
      senderName: json['sender_name'] as String,
      senderEmail: json['sender_email'] as String,
      subject: json['subject'] as String,
      body: json['body'] as String,
      listingId: json['listing_id'] as String?,
      timestamp: DateTime.parse(json['timestamp'] as String),
      isRead: json['is_read'] as bool? ?? false,
      aiDraftResponse: json['ai_draft_response'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'source': source.name,
      'sender_name': senderName,
      'sender_email': senderEmail,
      'subject': subject,
      'body': body,
      'listing_id': listingId,
      'timestamp': timestamp.toIso8601String(),
      'is_read': isRead,
      'ai_draft_response': aiDraftResponse,
    };
  }

  Message copyWith({
    String? id,
    MessageSource? source,
    String? senderName,
    String? senderEmail,
    String? subject,
    String? body,
    String? listingId,
    DateTime? timestamp,
    bool? isRead,
    String? aiDraftResponse,
  }) {
    return Message(
      id: id ?? this.id,
      source: source ?? this.source,
      senderName: senderName ?? this.senderName,
      senderEmail: senderEmail ?? this.senderEmail,
      subject: subject ?? this.subject,
      body: body ?? this.body,
      listingId: listingId ?? this.listingId,
      timestamp: timestamp ?? this.timestamp,
      isRead: isRead ?? this.isRead,
      aiDraftResponse: aiDraftResponse ?? this.aiDraftResponse,
    );
  }

  static MessageSource _parseSource(String source) {
    switch (source.toLowerCase()) {
      case 'email':
        return MessageSource.email;
      case 'ebay':
        return MessageSource.ebay;
      case 'facebook':
        return MessageSource.facebook;
      case 'craigslist':
        return MessageSource.craigslist;
      case 'mercari':
        return MessageSource.mercari;
      case 'etsy':
        return MessageSource.etsy;
      case 'amazon':
        return MessageSource.amazon;
      default:
        return MessageSource.email;
    }
  }
}
